### PR TITLE
Preventing gpstart if there is an unknown configuration parameter in postgresql.conf.

### PR DESCRIPTION
**Pull request by foyzur**
If there is an unknown configuration parameter in postgresql.conf,
currently we silently print a log message and skip reading any
configuration parameter from postgresql.conf. This does not prevent
database startup. As some of our gucs depend on explicitly mentioned
values on certain platforms (e.g., gp_vmem_protect_limit is set to 8GB
explicitly in postgresql.conf, and without explicit mention, it defaults
to 0 on darwin platforms), this can result in assertion failure and
confusing dtm errors as QEs fail to start with 0 vmem protect limit (and
potentially other invalid guc values).

This change prevent database startup if there is any unknown
configuration parameter.

This is relevant to Github issue #1201. However, for issue #1201 we will
need another fix to correct bugbuster test so that it doesn't put
unrecognized parameter in the postgresql.conf.

### CODE REVIEW

File: src/backend/utils/misc/guc-file.l [hunk 1]
```diff
@@ -259,7 +262,10 @@ ProcessConfigFile(GucContext context)
 
 		if (!set_config_option(item->name, item->value, context,
 							   PGC_S_FILE, GUC_ACTION_SET, false))
+		{
```
Comment (vraghavan78) [inline src/backend/utils/misc/guc-file.l]: can you add a comment here

Comment (foyzur) [inline src/backend/utils/misc/guc-file.l]: Sure.

**Comment by hlinnaka** [comment]
PostgreSQL will refuse to start up, if there's an unknown GUC in postgresql.conf. That had been intentionally disabled in GPDB, back in 2009, by this commit (from the historical git repository):

```
commit 5fb81f63a00e81baee152ad42878b3601565f1ac
Author: waasf <a@b>
Date:   Tue Mar 24 21:41:30 2009 -0800

    Change GUC processing to warn when encountering bad GUC instead of aborting startup.
    
    [git-p4: depot-paths = "//cdb2/main/": change = 32974]
```

Let's revert that to the way it's done in the upstream, rather than invent a new way of refusing to start up on unrecognized options.

**Comment by foyzur** [comment]
@hlinnaka could you please clarify? Do you want me to revert to Florian's commit? Or do you want me to bring the latest changes from postgresql? The latter may require lot of work.

**Comment by hlinnaka** [comment]
After the changes, the (relevant) code should look the same as in the upstream, in PostgreSQL 8.3. You could do it by reverting that commit, or copying the relevant bits from REL8_3_STABLE. The end result should be the same.

**Comment by ascherbaum-pivotal** [comment]
According to Florian, this was deactivated because it was much more practical to let startup proceed to make the downtime window as short as possible.

**Comment by danielgustafsson** [comment]
+1 on moving to the code from upstream 8.3, failing on incorrect GUC makes perfect sense and it will make further merging with upstream easier.

**Comment by foyzur** [comment]
Thanks everyone for your valuable feedback! I will explore how to port from 8.3.

**Comment by foyzur** [comment]
@hlinnaka @danielgustafsson @ascherbaum-pivotal @vraghavan78 @karthijrk please take a look at the new change that adopts 8.3 approach.

**Comment by foyzur** [comment]
@hlinnaka @danielgustafsson @ascherbaum-pivotal I am pushing as I have two +1. But if you have concerns I can revert and implement any suggestion that you may have.

**Comment by hlinnaka** [comment]
@foyzur, commit 5fb81f63a00e81baee152ad42878b3601565f1ac contained other changes as well. Are they still required? If not, we should revert them as well.

**Pull request merged**